src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs

```diff
@@ -82,12 +91,77 @@ public static class AutofacRegistration
             .SingleInstance();
 
         // Shims for keyed service compatibility.
-        builder.RegisterServiceMiddlewareSource(new KeyedServiceMiddlewareSource());
         builder.RegisterSource<AnyKeyRegistrationSource>();
+        builder.ComponentRegistryBuilder.Registered += AddFromKeyedServiceParameterMiddleware;
 
         Register(builder, descriptors, lifetimeScopeTagForSingletons);
     }
 
+    /// <summary>
+    /// Inspect each component registration, and determine whether or not we can avoid adding the
+    /// <see cref="FromKeyedServicesAttribute"/> parameter to the resolve pipeline.
+    /// </summary>
+    private static void AddFromKeyedServiceParameterMiddleware(object? sender, ComponentRegisteredEventArgs e)
+    {
+        var needFromKeyedServiceParameter = false;
+
+        // We can optimise quite significantly in the case where we are using the reflection activator.
+        // In that state we can inspect the constructors ahead of time and determine whether the parameter will even need to be added.
+        if (e.ComponentRegistration.Activator is ReflectionActivator reflectionActivator)
+        {
+            var constructors = reflectionActivator.ConstructorFinder.FindConstructors(reflectionActivator.LimitType);
+
+            // Go through all the constructors; if any have a FromKeyedServices, then we must add our component middleware to
+            // the pipeline.
+            foreach (var constructor in constructors)
+            {
+                foreach (var constructorParameter in constructor.GetParameters())
+                {
+                    if (constructorParameter.GetCustomAttribute<FromKeyedServicesAttribute>() is not null)
+                    {
+                        // One or more of the constructors we will use to activate has a FromKeyedServicesAttribute,
+                        // we must add our middleware.
+                        needFromKeyedServiceParameter = true;
+                        break;
+                    }
+                }
+
+                if (needFromKeyedServiceParameter)
+                {
+                    break;
+                }
+            }
+        }
+        else if (e.ComponentRegistration.Activator is DelegateActivator)
+        {
+            // With non-reflection activation, there are very few paths that would let the FromKeyedServicesAttribute
+            // actually work.
+            needFromKeyedServiceParameter = false;
```

Comment:
The only case I can find where someone might reasonably expect `FromKeyedServices` to be respected on a delegate activation is this one:

```
builder.Register(([FromKeyedServices("abc")] KeyedService service) => new ServiceConsumer(service));
```

Note that that's obviously not an MSDI registration, but it does kind of feel like something that _should_ work?

If we need to support that case, then we will have to add our middleware to all delegate activations, meaning that delegate activation slows down slightly.

If we don't need to support that case, then we can leave it false and never response `[FromKeyedServices]` on a delegate activation.

Reply:
I'd say, for now, to skip this case and let delegates work fast. It's a pretty edge case thing, and it's absolutely doing the whole mix-and-match MEDI and Autofac registration types that we pretty clearly tell people _not_ to do. I didn't figure out a simple way to replicate anything like this in straight MEDI, so if mix-and-match doesn't work? I'm OK with that, at least for this.